pkg/agent/cniserver/pod_monitor.go

```diff
@@ -0,0 +1,208 @@
+	unreadyInterfaces map[string]*unreadyPodInfo
+	// mu is a mutex to guard map unreadyInterfaces.
+	mu       sync.Mutex
+	statusCh chan *openflow15.PortStatus
+}
+
+func (m *podIfaceMonitor) Run(stopCh <-chan struct{}) {
+	klog.Info("Starting the monitor to wait for new OpenFlow ports")
+	m.eventBroadcaster.StartStructuredLogging(0)
+	m.eventBroadcaster.StartRecordingToSink(&v1.EventSinkImpl{
+		Interface: m.kubeClient.CoreV1().Events(""),
+	})
+	defer m.eventBroadcaster.Shutdown()
+
+	go func() {
+		ticker := time.Tick(retryInterval)
+		for {
+			select {
+			case <-stopCh:
+				return
+			case status := <-m.statusCh:
+				klog.V(2).InfoS("Received PortStatus message", "message", status)
+				// Update Pod OpenFlow entries only after the OpenFlow port state is live.
+				if status.Desc.State == openflow15.PS_LIVE {
+					m.updateUnreadyPod(status)
+				}
+			case <-ticker:
+				m.checkUnreadyPods()
+			}
+		}
+	}()
+}
+
+func (m *podIfaceMonitor) updatePodFlows(ifName string, ofPort int32, podInfo *unreadyPodInfo) error {
+	ifConfig, found := m.ifaceStore.GetInterfaceByName(ifName)
+	if !found {
+		klog.InfoS("Interface config is not found in local cache, remove from local cache", "port", ifName)
+		delete(m.unreadyInterfaces, ifName)
+		return nil
+	}
+
+	// Update interface config with the ofPort.
+	ifConfig.OVSPortConfig.OFPort = ofPort
+	m.ifaceStore.UpdateInterface(ifConfig)
+
+	// Install OpenFlow entries for the Pod.
+	klog.V(2).InfoS("Setting up Openflow entries for OVS port", "port", ifName)
+	if err := m.ofClient.InstallPodFlows(ifName, ifConfig.IPs, ifConfig.MAC, uint32(ofPort), ifConfig.VLANID, nil); err != nil {
+		return fmt.Errorf("failed to add Openflow entries for OVS port %s: %v", ifName, err)
+	}
+
+	// Notify the Pod update event to required components.
+	event := agenttypes.PodUpdate{
+		PodName:      ifConfig.PodName,
+		PodNamespace: ifConfig.PodNamespace,
+		IsAdd:        true,
+		ContainerID:  ifConfig.ContainerID,
+	}
+	m.podUpdateNotifier.Notify(event)
+
+	podInfo.flowInstalled = true
+	// Update Pod event immediately
+	m.processPodEvents(ifName, podInfo)
+	return nil
+}
+
+func (m *podIfaceMonitor) updateUnreadyPod(status *openflow15.PortStatus) {
+	ovsPort := string(bytes.Trim(status.Desc.Name, "\x00"))
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	podInfo, found := m.unreadyInterfaces[ovsPort]
+	if !found {
+		klog.InfoS("OVS port is not found", "ovsPort", ovsPort)
+		return
+	}
+
+	ofPort := status.Desc.PortNo
+	if err := m.updatePodFlows(ovsPort, int32(ofPort), podInfo); err != nil {
+		klog.ErrorS(err, "Failed to update Pod's OpenFlow entries, requeue the PortStatus message after 10s", "PodName", podInfo.podName, "PodNamespace", podInfo.podNamespace, "OVSPort", ovsPort)
+		m.reQueuePortStatus(status)
+		return
+	}
+}
+
+func (m *podIfaceMonitor) checkUnreadyPods() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	for key, podInfo := range m.unreadyInterfaces {
+		m.processPodEvents(key, podInfo)
+	}
```

Comment:
typically I would have said it's not a great idea to hold the lock while doing expensive tasks, but it seems that generating events is non-blocking (event is added to queue), so this should be fine as is

Reply:
I agree it is not a good design to hold the lock outside the for loop. I choose this solution is because we use the lock on the map level, so we have to lock it when we loop the items inside the map.

I used to think about using RWMutex to reduce the lock time, e.g., use RLock when reading/looping the items in the map, and use Lock when we add/delete items in the map. Then a challenge I need to handle is I can't use WLock inside a RLock, which causes a deadlock, then I have to split the check-delete into two steps, 1) find the "is-ready" Pods inside the map using a RLock and update events, then 2) delete them within another Lock after the RLock is released. Besides, we also need to think about how to share code with the worker logic in function `updatePodFlows` which needs to check the existence of the OVS port before installing OpenFlow entries, and update the Pod event as ready after flows are installed successfully.So I still feel it is not good in the code if change to RWLock as it seems not reduce much cost.

Could you share your thought to refine this code?

Reply:
It seems to me that the ticker and hence the `checkUnreadyPods` function are no longer needed with the workqueue design. With the workqueue, a Pod will be kept in the workqueue until its network is ready or until the Pod no longer exists. We use the following enqueue operations for the Pod (or rather the corresponding port):
- when the Pod is first added, we use a 30s delay (timeout to declare the Pod not ready) with `AddAfter`
- when a port status is received, we `Add` the Pod right away so that the status update can be processed in a timely fashion
- when a Pod is processed (by `processNextWorkItem`), if the Pod is not ready yet, we add the Pod again with `AddAfter` and a delay of 5s

With this design, only the workqueue worker is responsible for all updates and we only handle one Pod / port at a time. Right now we have 2 goroutines involved (worker and ticker), but the ticker is not needed. That will probably simplify the locking scheme as well.

Reply:
It sounds great to me, updated as suggested.